rust/otap-dataflow/crates/otlp/src/otlp_batch_processor.rs

```diff
@@ -13,6 +13,113 @@ use prost::Message as ProstMessage;
 use std::borrow::Cow;
 use std::time::{Duration, Instant};
 
+/// Trait for a batch type (e.g., ExportTraceServiceRequest, ExportMetricsServiceRequest, ExportLogsServiceRequest)
+pub trait Batch: Sized {
+    type Resource: ResourceGroup;
+    fn resources_mut(&mut self) -> &mut Vec<Self::Resource>;
+    fn new_empty() -> Self;
+}
+
+/// Trait for a resource group (e.g., ResourceSpans, ResourceMetrics, ResourceLogs)
+pub trait ResourceGroup: Sized {
+    type Scope: ScopeGroup;
+    fn scopes_mut(&mut self) -> &mut Vec<Self::Scope>;
+    fn take_resource_fields(&mut self) -> Self;
+}
+
+/// Trait for a scope group (e.g., ScopeSpans, ScopeMetrics, ScopeLogs)
+pub trait ScopeGroup: Sized {
+    type Leaf;
+    fn leaves_mut(&mut self) -> &mut Vec<Self::Leaf>;
+    fn take_scope_fields(&mut self) -> Self;
+}
+
+/// -- Traces --
+impl Batch for ExportTraceServiceRequest {
+    type Resource = ResourceSpans;
+    fn resources_mut(&mut self) -> &mut Vec<Self::Resource> {
+        &mut self.resource_spans
+    }
+    fn new_empty() -> Self {
+        ExportTraceServiceRequest {
+            resource_spans: Vec::new(),
+        }
+    }
+}
+
+impl ResourceGroup for ResourceSpans {
+    type Scope = ScopeSpans;
+    fn scopes_mut(&mut self) -> &mut Vec<Self::Scope> {
+        &mut self.scope_spans
+    }
+    fn take_resource_fields(&mut self) -> Self {
+        ResourceSpans {
+            resource: self.resource.take(),
+            scope_spans: Vec::new(),
+            schema_url: self.schema_url.clone(),
+        }
+    }
+}
+
+impl ScopeGroup for ScopeSpans {
+    type Leaf = crate::proto::opentelemetry::trace::v1::Span;
+    fn leaves_mut(&mut self) -> &mut Vec<Self::Leaf> {
+        &mut self.spans
+    }
+    fn take_scope_fields(&mut self) -> Self {
+        ScopeSpans {
+            scope: self.scope.clone(),
```

Comment:
Should this be take()? I noticed that ResourceSpans uses take() in this context.